README.md
# docker-gitea

[![Build Status](https://travis-ci.com/madharjan/docker-gitea.svg?branch=master)](https://travis-ci.com/madharjan/docker-gitea)
[![Layers](https://images.microbadger.com/badges/image/madharjan/docker-gitea.svg)](http://microbadger.com/images/madharjan/docker-gitea)

Docker container for Gitea Server based on [madharjan/docker-base](https://github.com/madharjan/docker-base/)

## Features

* Environment variables to set database setting or link to postgresql container
* Environment variables to configure Gitea
* Bats [bats-core/bats-core](https://github.com/bats-core/bats-core) based test cases

## Gitea 1.8.2 (docker-gitea)

### Environment

| Variable                   | Default                | Example        |
| -------------------------- | ---------------------- | -------------- |
| DISABLE_GITEA              | 0                      | 1 (to disable) |
|                            |                        |                |
| GITEA_PROTOCOL             | http                   | https          |
| GITEA_DOMAIN               | localhost              |                |
|                            |                        |                |
| GITEA_SMTP_HOST            | localhost              |                |
| GITEA_SMTP_PORT            | 25                     |                |
| GITEA_SMTP_FROM            | gitea@localhost        |                |
| GITEA_SMTP_USER            | pass                   |                |
| GITEA_SMTP_PASS            | pass                   |                |
|                            |                        |                |
| GITEA_REGISTRATION_DISABLE | false                  |                |
| GITEA_RECAPTCHA_SECRET     |                        |                |
| GITEA_RECAPTCHA_KEY        |                        |                |
|                            |                        |                |
| GITEA_INSTALL_LOCK         | false                  |                |
| GITEA_SECRET_KEY           |                        |                |
| GITEA_JWT_SECRET           |                        |                |
|                            |                        |                |
| GITEA_ROOT_EMAIL           | root@localhost         |                |
| GITEA_ROOT_PASSWORD        | Git3aPa55              |                |
|                            |                        |                |
| GITEA_DB_TYPE              | postgres               |                |
| GITEA_DB_HOST              | linked to 'postgresql' | 172.17.0.1     |
| GITEA_DB_PORT              | linked to 'postgresql' | 5432           |
| GITEA_DB_USER              | linked to 'postgresql' | gitea          |
| GITEA_DB_PASS              | linked to 'postgresql' | gitea          |

## Build

```bash
# clone project
git clone https://github.com/madharjan/docker-gitea
cd docker-gitea

# build
make

# tests
make run
make test

# clean
make clean
```

## Run

**Note**: update environment variables below as necessary

```bash
# prepare foldor on host for container volumes
sudo mkdir -p /opt/docker/gitea/postgresql/etc/
sudo mkdir -p /opt/docker/gitea/postgresql/lib/
sudo mkdir -p /opt/docker/gitea/postgresql/log/

# stop & remove previous instances
docker stop gitea-postgresql
docker rm gitea-postgresql

# run container
docker run -d \
  -e POSTGRESQL_DATABASE=gitea \
  -e POSTGRESQL_USERNAME=gitea \
  -e POSTGRESQL_PASSWORD=Pa55w0rd \
  -v /opt/docker/gitea/postgresql/etc:/etc/postgresql/9.5/main \
  -v /opt/docker/gitea/postgresql/lib:/var/lib/postgresql/9.5/main \
  -v /opt/docker/gitea/postgresql/log:/var/log/postgresql \
  --name gitea-postgresql \
  madharjan/docker-postgresql:9.5

# prepare foldor on host for container volumes
sudo mkdir -p /opt/docker/gitea/etc/
sudo mkdir -p /opt/docker/gitea/lib/
sudo mkdir -p /opt/docker/gitea/log/

# stop & remove previous instances
docker stop gitea
docker rm gitea

# run container linked with gitea-postgresql
docker run -d \
  --link gitea-postgresql:postgresql \
  -e GITEA_INSTALL_LOCK=true \
  -e GITEA_SECRET_KEY=1234567890 \
  -e GITEA_JWT_SECRET=6Lfjq6YUAAAAAFwDmDtfHyHmL1234567890 \
  -e GITEA_ROOT_PASSWORD=Pa55w0rd \
  -p 3000:3000 \
  -v /opt/docker/gitea/etc:/etc/gitea \
  -v /opt/docker/gitea/lib:/var/lib/gitea \
  -v /opt/docker/gitea/log:/var/log/gitea \
  --name gitea \
  madharjan/docker-gitea:1.8.2
```

## Systemd Unit File

**Note**: update environment variables below as necessary

```txt
[Unit]
Description=Gitea Server

After=docker.service

[Service]
TimeoutStartSec=0

ExecStartPre=-/bin/mkdir -p /opt/docker/gitea/etc
ExecStartPre=-/bin/mkdir -p /opt/docker/gitea/lib
ExecStartPre=-/bin/mkdir -p /opt/docker/gitea/log
ExecStartPre=-/usr/bin/docker stop gitea
ExecStartPre=-/usr/bin/docker rm gitea
ExecStartPre=-/usr/bin/docker pull madharjan/docker-gitea:9.5

ExecStart=/usr/bin/docker run \
  --link odoo-postgresql:postgresql \
  -e GITEA_INSTALL_LOCK=true \
  -e GITEA_SECRET_KEY=1234567890 \
  -e GITEA_JWT_SECRET=6Lfjq6YUAAAAAFwDmDtfHyHmL1234567890 \
  -e GITEA_ROOT_PASSWORD=Pa55w0rd \
  -p 80:3000 \
  -v /opt/docker/gitea/etc:/etc/gitea \
  -v /opt/docker/gitea/lib:/var/lib/gitea \
  -v /opt/docker/gitea/log:/var/log/gitea \
  --name gitea \
  madharjan/docker-gitea:1.0

ExecStop=/usr/bin/docker stop -t 2 gitea

[Install]
WantedBy=multi-user.target
```

## Generate Systemd Unit File

| Variable                   | Default     | Example              |
| -------------------------- | ----------- | -------------------- |
| NAME                       | gitea       |                      |
| VOLUME_HOME                | /opt/docker | /opt/data            |
| PORT                       |             | 8080                 |
| LINK_CONTAINERS            |             | postgresql           |
|                            |             |                      |
| GITEA_PROTOCOL             |             | https                |
| GITEA_DOMAIN               |             | git.company.com      |
|                            |             |                      |
| GITEA_SMTP_HOST            |             | mail.company.com     |
| GITEA_SMTP_PORT            |             | 25                   |
| GITEA_SMTP_FROM            |             | gitea@company.com    |
| GITEA_SMTP_USER            |             |                      |
| GITEA_SMTP_PASS            |             |                      |
|                            |             |                      |
| GITEA_REGISTRATION_DISABLE |             | true                 |
| GITEA_RECAPTCHA_SECRET     |             | 12345678900987654321 |
| GITEA_RECAPTCHA_KEY        |             |                      |
|                            |             |                      |
| GITEA_INSTALL_LOCK         |             | true                 |
| GITEA_SECRET_KEY           |             | 12345678900987654321 |
| GITEA_JWT_SECRET           |             | 12345678900987654321 |
|                            |             |                      |
| GITEA_ROOT_EMAIL           |             | root@company.com     |
| GITEA_ROOT_PASSWORD        |             | Pa55w0rd             |

```bash
# generate postgresql.service
docker run --rm \
  -e NAME=gitea-postgresql \
  -e POSTGRESQL_DATABASE=gitea \
  -e POSTGRESQL_USERNAME=gitea \
  -e POSTGRESQL_PASSWORD=Pa55w0rd \
  madharjan/docker-postgresql:9.5 \
  postgresql-systemd-unit | \
  sudo tee /etc/systemd/system/gitea-postgresql.service

sudo systemctl enable gitea-postgresql
sudo systemctl start gitea-postgresql

# generate gitea.service
docker run --rm \
  -e PORT=80 \
  -e LINK_CONTAINERS=gitea-postgresql:postgresql,mail:mail \
  -e GITEA_INSTALL_LOCK=true \
  -e GITEA_SECRET_KEY=1234567890 \
  -e GITEA_JWT_SECRET=6Lfjq6YUAAAAAFwDmDtfHyHmL1234567890 \
  -e GITEA_ROOT_PASSWORD=Pa55w0rd \
  -e GITEA_SMTP_HOST=mail \
  -e GITEA_SMTP_FROM=gitea@company.com \
  madharjan/docker-gitea:1.8.2 \
  gitea-systemd-unit | \
  sudo tee /etc/systemd/system/gitea.service

sudo systemctl enable gitea
sudo systemctl start gitea
```


Changelog.md
## 1.8.2

### Changes

* Based on madharjan/docker-base:16.04
* Using Gitea 1.8.2


test/run.sh
#!/bin/bash

set -x

DIR=$(dirname $0)
$DIR/clean.sh

sudo mkdir -p /opt/docker/gitea/postgresql/etc/
sudo mkdir -p /opt/docker/gitea/postgresql/lib/
sudo mkdir -p /opt/docker/gitea/postgresql/log/

docker run -d -t \
  -e DEBUG=false \
  -e POSTGRESQL_DATABASE=gitea \
  -e POSTGRESQL_USERNAME=gitea \
  -e POSTGRESQL_PASSWORD=Pa55w0rd \
  -v /opt/docker/gitea/postgresql/etc:/etc/postgresql/9.5/main \
  -v /opt/docker/gitea/postgresql/lib:/var/lib/postgresql/9.5/main \
  -v /opt/docker/gitea/postgresql/log:/var/log/postgresql \
  --name gitea-postgresql \
  madharjan/docker-postgresql:9.5

sleep 3
#docker logs gitea-postgresql

sudo mkdir -p /opt/docker/gitea/etc/
sudo mkdir -p /opt/docker/gitea/lib/
sudo mkdir -p /opt/docker/gitea/log/

docker run -d -t \
  -e DEBUG=true \
  --link gitea-postgresql:postgresql \
  -e GITEA_PROTOCOL=http \
  -e GITEA_DOMAIN=192.168.56.254 \
  -e GITEA_SMTP_PORT=2525 \
  -e GITEA_REGISTRATION_DISABLE=false \
  -e GITEA_INSTALL_LOCK=true \
  -e GITEA_SECRET_KEY=1234567890 \
  -e GITEA_JWT_SECRET=6Lfjq6YUAAAAAFwDmDtfHyHmL1234567890 \
  -e GITEA_ROOT_EMAIL=root@local.host \
  -e GITEA_ROOT_PASSWORD=Pa55w0rd \
  -p 80:3000 \
  -v /opt/docker/gitea/etc:/etc/gitea \
  -v /opt/docker/gitea/lib:/var/lib/gitea \
  -v /opt/docker/gitea/log:/var/log/gitea \
  --name gitea \
  madharjan/docker-gitea:1.8.2

sleep 2
docker logs -f gitea


test/clean.sh
#!/bin/bash

set -x

docker stop gitea 2> /dev/null
docker rm gitea 2> /dev/null

docker stop gitea-postgresql 2> /dev/null
docker rm gitea-postgresql 2> /dev/null

sudo rm -rf /opt/docker/gitea


services/gitea/gitea.sh
#!/bin/bash
set -e
export LC_ALL=C
export DEBIAN_FRONTEND=noninteractive

if [ "${DEBUG}" = true ]; then
  set -x
fi

GITEA_BUILD_PATH=/build/services/gitea
GITEA_VERSION=1.8.2

## Install Nginx.
apt-get install -y --no-install-recommends git-core iproute2

## Install Gitea Server
wget -nv "https://github.com/go-gitea/gitea/releases/download/v${GITEA_VERSION}/gitea-${GITEA_VERSION}-linux-amd64" -O /usr/local/bin/gitea
#wget -nv "https://dl.gitea.io/gitea/${GITEA_VERSION}/gitea-${GITEA_VERSION}-linux-amd64" -O /usr/local/bin/gitea
chmod +x /usr/local/bin/gitea

## Create 'git' user/group
addgroup --gid 1000 git 
adduser --system \
    --home /home/git \
    --shell /bin/bash \
    --uid 1000 \
    --gid 1000 \
    git

mkdir -p /etc/service/gitea
cp ${GITEA_BUILD_PATH}/gitea.runit /etc/service/gitea/run
chmod 750 /etc/service/gitea/run

## Configure logrotate


services/20-gitea.sh
#!/bin/bash

set -e

if [ "${DEBUG}" = true ]; then
  set -x
fi

function waitAandInitializeGitea()
{
  sleep 4
  date
  echo "Waiting for Gitea Server to come up ..."

  while read LINE; \
  do \
    if [[ $LINE =~ .*Listen.* ]]; then \
      echo "Gitea Server is up"; \
      break; \
    fi \
  done < <(tail -f /var/log/gitea/gitea.log)

  echo "Configuring Gitea ...."
  export GITEA_WORK_DIR=/var/lib/gitea
  cd /home/git
  /sbin/setuser git /usr/local/bin/gitea admin create-user -c /etc/gitea/app.ini --name root --password ${GITEA_ROOT_PASSWORD} --email ${GITEA_ROOT_EMAIL} --admin
  echo "Configuring Gitea Done"
  date
  exit 0
}

DISABLE_GITEA=${DISABLE_GITEA:-0}

GITEA_DB_TYPE=${GITEA_DB_TYPE:-postgres}
GITEA_DB_HOST=${GITEA_DB_HOST:-${POSTGRESQL_PORT_5432_TCP_ADDR}}
if [ ${GITEA_DB_PORT} =~ ^[-0-9]+$ ]; then
  GITEA_DB_PORT=${POSTGRESQL_PORT}
else
  GITEA_DB_PORT=${POSTGRESQL_PORT_5432_TCP_PORT}
fi
GITEA_DB_NAME=${GITEA_DB_NAME:-$POSTGRESQL_ENV_POSTGRESQL_DATABASE}
GITEA_DB_USER=${GITEA_DB_USER:-$POSTGRESQL_ENV_POSTGRESQL_USERNAME}
GITEA_DB_PASS=${GITEA_DB_PASS:-$POSTGRESQL_ENV_POSTGRESQL_PASSWORD}

GITEA_PROTOCOL=${GITEA_PROTOCOL:-http}
GITEA_DOMAIN=${GITEA_DOMAIN:-localhost}

DEF_GITEA_SMTP_HOST=$(ip route | grep default | awk '{print$3}')
GITEA_SMTP_HOST=${GITEA_SMTP_HOST:-$DEF_GITEA_SMTP_HOST}
GITEA_SMTP_PORT=${GITEA_SMTP_PORT:-25}
GITEA_SMTP_FROM=${GITEA_SMTP_FROM:-gitea@localhost}
GITEA_SMTP_USER=${GITEA_SMTP_USER:-}
GITEA_SMTP_PASS=${GITEA_SMTP_USER:-}

GITEA_REGISTRATION_DISABLE=${GITEA_REGISTRATION_DISABLE:-false}
GITEA_RECAPTCHA_SECRET=${GITEA_RECAPTCHA_SECRET}
GITEA_RECAPTCHA_KEY=${GITEA_RECAPTCHA_KEY}

GITEA_INSTALL_LOCK=${GITEA_INSTALL_LOCK:-false}
GITEA_SECRET_KEY=${GITEA_SECRET_KEY:-}
GITEA_JWT_SECRET=${GITEA_JWT_SECRET:-}

GITEA_ROOT_EMAIL=${GITEA_ROOT_EMAIL:-root@localhost}
GITEA_ROOT_PASSWORD=${GITEA_ROOT_PASSWORD:-Git3aPa55}

GITEA_SMTP_ENABLE=${GITEA_SMTP_HOST:+true}
GITEA_RECAPTCHA_ENABLE=${GITEA_RECAPTCHA_SECRET:+true}

mkdir -p /var/lib/gitea/{custom,data,tmp,static}
mkdir -p /etc/gitea
mkdir -p /var/log/gitea


if [ ! "${DISABLE_GITEA}" -eq 0 ]; then
  touch /etc/service/gitea/down
else
  rm -f /etc/service/gitea/down
fi

if [ -f /etc/gitea/app.ini ]; then
  echo "Gitea config 'app.ini' already exists"
else
  cp /config/etc/gitea/app.ini /etc/gitea/app.ini
  export GITEA_DB_TYPE GITEA_DB_HOST GITEA_DB_NAME GITEA_DB_USER GITEA_DB_PASS
  export GITEA_PROTOCOL GITEA_DOMAIN 
  export GITEA_SMTP_ENABLE GITEA_SMTP_HOST GITEA_SMTP_PORT GITEA_SMTP_FROM GITEA_SMTP_USER GITEA_SMTP_PASS
  export GITEA_RECAPTCHA_ENABLE GITEA_RECAPTCHA_SECRET GITEA_RECAPTCHA_KEY GITEA_REGISTRATION_DISABLE
  export GITEA_INSTALL_LOCK GITEA_SECRET_KEY
  perl -p -i -e 's/\$\{([^}]+)\}/defined $ENV{$1} ? $ENV{$1} : $&/eg' /etc/gitea/app.ini
 
  #Background the function call
  waitAandInitializeGitea &
fi

mkdir -p /var/lib/gitea/custom/templates

if [ -f /var/lib/gitea/custom/templates/home.tmpl ]; then
  echo "Gitea templates already exists"
else
  cp -r /config/templates/* /var/lib/gitea/custom/templates
fi

chown -R git:git /var/lib/gitea/
chmod -R 750 /var/lib/gitea/
chown -R root:git /etc/gitea
chmod -R 770 /etc/gitea
chown -R root:git /var/log/gitea
chmod -R 770 /var/log/gitea



